services/user/Subgroups/service/src/lib.rs

```diff
@@ -0,0 +1,192 @@
+#[psibase::service(name = "subgroups")]
+pub mod service {
+    pub use psibase::fracpack;
+    use std::collections::HashMap;
+
+    /// # Description
+    ///
+    /// Recursive function that finds a valid partition of `n` into groups with sizes from `allowed`
+    /// Uses memoization to avoid redundant calculations
```

Comment:
dynamic programming is more efficient

Reply:
Not sure what you're looking for, but I would propose if efficiency is the only concern that we proceed since this has the desired properties, and we can optimize more later

Reply:
I suspect it would also be simpler:

```rust
for i in 0..=n {
  groups[i] = allowed.iter().map(|g| append_solution(&groups[i - g], g)).max_by(choose_best_solution);
}
```

Reply:
Hmm, okay, I think I understand and agree it's better. I'll give it a shot

Reply:
Okay, that ended up being an easier refactor than I thought. I like this implementation much more.